pkg/controller/configmapcainjector/configmap_injecting_controller.go

```diff
@@ -0,0 +1,178 @@
+	configMapClient kcoreclient.ConfigMapsGetter
+
+	// configMaps that need to be checked
+	queue workqueue.RateLimitingInterface
+
+	configMapLister    listers.ConfigMapLister
+	configMapHasSynced cache.InformerSynced
+
+	ca string
+
+	// syncHandler does the work. It's factored out for unit testing
+	syncHandler func(serviceKey string) error
+}
+
+// NewConfigMapCABundleInjectionController creates a new ServiceServingCertUpdateController.
+// TODO this should accept a shared informer
+func NewConfigMapCABundleInjectionController(configMaps informers.ConfigMapInformer, configMapsClient kcoreclient.ConfigMapsGetter, ca []byte, resyncInterval time.Duration) *ConfigMapCABundleInjectionController {
+	ic := &ConfigMapCABundleInjectionController{
+		queue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		ca:    string(ca),
+	}
+
+	ic.configMapLister = configMaps.Lister()
+	configMaps.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    ic.addConfigMap,
+			UpdateFunc: ic.updateConfigMap,
+		},
+		resyncInterval,
+	)
+	ic.configMapHasSynced = configMaps.Informer().GetController().HasSynced
+	ic.configMapClient = configMapsClient
+	ic.syncHandler = ic.syncConfigMap
+
+	return ic
+}
+
+// Run begins watching and syncing.
+func (ic *ConfigMapCABundleInjectionController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer ic.queue.ShutDown()
+
+	// Wait for the stores to fill
+	if !cache.WaitForCacheSync(stopCh, ic.configMapHasSynced) {
+		return
+	}
+
+	glog.V(1).Infof("Starting workers")
+	for i := 0; i < workers; i++ {
+		go wait.Until(ic.runWorker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.V(1).Infof("Shutting down")
+}
+
+func (ic *ConfigMapCABundleInjectionController) enqueueConfigMap(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+
+	ic.queue.Add(key)
+}
+
+func (ic *ConfigMapCABundleInjectionController) addConfigMap(obj interface{}) {
+	cm, ok := obj.(*v1.ConfigMap)
+	if !ok {
+		return
+	}
+
+	if !hasCABundleAnnotation(cm) {
+		return
+	}
+
+	glog.V(4).Infof("adding %s", cm.Name)
+	ic.enqueueConfigMap(cm)
+}
+
+func (ic *ConfigMapCABundleInjectionController) updateConfigMap(old, cur interface{}) {
+	cm, ok := cur.(*v1.ConfigMap)
+	if !ok {
+		return
+	}
+
+	if !hasCABundleAnnotation(cm) {
+		return
+	}
+
+	glog.V(4).Infof("updating %s", cm.Name)
+	ic.enqueueConfigMap(cm)
+}
+
+func (ic *ConfigMapCABundleInjectionController) runWorker() {
+	for ic.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (ic *ConfigMapCABundleInjectionController) processNextWorkItem() bool {
+	key, quit := ic.queue.Get()
+	if quit {
+		return false
+	}
+	defer ic.queue.Done(key)
+
+	err := ic.syncHandler(key.(string))
+	if err == nil {
```

Comment:
nit: stay consistent with handling error cases in the `if`s, with the final return of the function handling the default success case. i.e. switch this to `if err != nil {` and swap remaining blocks